tests fail on 32bit and BigEndian

# 32bit

```rust
     Running `/usr/bin/rustc --crate-name parse_self tests/parse_self.rs --emit=dep-info,link -C opt-level=3 --test -C metadata=f2a679e3f001d097 -C extra-filename=-f2a679e3f001d097 --out-dir /builddir/build/BUILD/gimli-0.15.0/target/release/deps -L dependency=/builddir/build/BUILD/gimli-0.15.0/target/release/deps --extern test_assembler=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libtest_assembler-5ef368c74a498911.rlib --extern object=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libobject-a864633da8d9ccb2.rlib --extern byteorder=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libbyteorder-3e5c4aca9ed64de9.rlib --extern getopts=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libgetopts-d69514fb74ea2875.rlib --extern memmap=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libmemmap-10d0e94555babc08.rlib --extern fallible_iterator=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libfallible_iterator-63284fa240e8f0d2.rlib --extern arrayvec=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libarrayvec-b0d3c97f2439cf98.rlib --extern gimli=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libgimli-0425f9e292cdcc61.rlib -Copt-level=3 -Cdebuginfo=2 -Clink-arg=-Wl,-z,relro,-z,now`
error[E0425]: cannot find value `input` in this scope
   --> src/parser.rs:592:15
    |
592 |         match input.read_offset(Format::Dwarf64) {
    |               ^^^^^ not found in this scope
warning: unused import: `parse_type_unit_header`
    --> src/unit.rs:2806:17
     |
2806 |                 parse_type_unit_header, parse_unit_header};
     |                 ^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: #[warn(unused_imports)] on by default
warning: unused import: `loc::DebugLocOffset`
    --> src/unit.rs:2813:9
     |
2813 |     use loc::DebugLocOffset;
     |         ^^^^^^^^^^^^^^^^^^^
error[E0599]: no associated item named `MAX` found for type `u32` in the current scope
   --> src/abbrev.rs:564:26
    |
564 |         let wrap_code = (u32::MAX as u64 + 1) + 1;
    |                          ^^^^^^^^
```

# BigEndian

```rust
     Running `/usr/bin/rustc --crate-name gimli src/lib.rs --emit=dep-info,link -C opt-level=3 --test -C metadata=58e0029c4aa85f0a -C extra-filename=-58e0029c4aa85f0a --out-dir /builddir/build/BUILD/gimli-0.15.0/target/release/deps -L dependency=/builddir/build/BUILD/gimli-0.15.0/target/release/deps --extern memmap=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libmemmap-ec02532aebe8cb19.rlib --extern test_assembler=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libtest_assembler-3e922a7b909d8031.rlib --extern object=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libobject-ed1c40f3e32b3419.rlib --extern byteorder=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libbyteorder-acd3dc1e025ae932.rlib --extern getopts=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libgetopts-c04bfc2976840f15.rlib --extern fallible_iterator=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libfallible_iterator-1717ebb2559f88d3.rlib --extern arrayvec=/builddir/build/BUILD/gimli-0.15.0/target/release/deps/libarrayvec-735c2575e8f4bf70.rlib -Copt-level=3 -Cdebuginfo=2 -Clink-arg=-Wl,-z,relro,-z,now`
error[E0308]: mismatched types
   --> src/parser.rs:606:9
    |
606 |         assert_eq!(*input, EndianBuf::new(&[1, 2, 3, 4], LittleEndian));
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `endianity::BigEndian`, found struct `endianity::LittleEndian`
    |
    = note: expected type `endianity::EndianBuf<'_, endianity::BigEndian>`
               found type `endianity::EndianBuf<'_, endianity::LittleEndian>`
    = note: this error originates in a macro outside of the current crate
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -28,7 +28,7 @@ before_script:
 - echo PATH is $PATH
 - export RUST_BACKTRACE=1
 
-script: . ./ci/script.sh
+script: ./ci/script.sh
 
 env:
   matrix:
@@ -56,3 +56,19 @@ matrix:
       env: GIMLI_JOB="alloc"        GIMLI_PROFILE=
     - rust: nightly
       env: GIMLI_JOB="alloc"        GIMLI_PROFILE="--release"
+    # Build a 32 bit target.
+    - rust: stable
+      sudo: required
+      services: docker
+      env:
+        - TARGET=i686-unknown-linux-gnu
+        - GIMLI_JOB=cross
+        - GIMLI_PROFILE=
+    # Build a big-endian target.
+    - rust: stable
+      sudo: required
+      services: docker
+      env:
+        - TARGET=powerpc64-unknown-linux-gnu
+        - GIMLI_JOB=cross
+        - GIMLI_PROFILE=

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -4,7 +4,6 @@ set -ex
 
 case "$GIMLI_JOB" in
     "test")
-        cargo clean
         cargo build $GIMLI_PROFILE
         cargo test $GIMLI_PROFILE
         ;;
@@ -31,7 +30,6 @@ case "$GIMLI_JOB" in
 
     "alloc")
         test "$TRAVIS_RUST_VERSION" == "nightly"
-        cargo clean
         cargo build --no-default-features --features alloc $GIMLI_PROFILE
         ;;
 
@@ -42,6 +40,12 @@ case "$GIMLI_JOB" in
     "coverage")
         bash <(curl "https://raw.githubusercontent.com/xd009642/tarpaulin/master/travis-install.sh");
         cargo tarpaulin --verbose --no-count --ciserver travis-ci --coveralls "$TRAVIS_JOB_ID";
+        ;;
+
+    "cross")
+        rustup target add $TARGET
+        cargo install cross --force
+        cross test --target $TARGET $GIMLI_PROFILE
         ;;
 
     *)

--- a/src/abbrev.rs
+++ b/src/abbrev.rs
@@ -393,7 +393,8 @@ pub mod tests {
     use endianity::{EndianBuf, LittleEndian};
     use parser::Error;
     use self::test_assembler::Section;
-    use std::u64;
+    #[cfg(target_pointer_width = "32")]
+    use std::u32;
     use test_util::GimliSectionMethods;
 
     pub trait AbbrevSectionMethods {

--- a/src/cfi.rs
+++ b/src/cfi.rs
@@ -6048,6 +6048,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg(target_pointer_width = "64")]
     fn size_of_unwind_ctx() {
         use std::mem;
         assert_eq!(
@@ -6059,6 +6060,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg(target_pointer_width = "64")]
     fn size_of_register_rule_map() {
         use std::mem;
         assert_eq!(

--- a/src/op.rs
+++ b/src/op.rs
@@ -1788,8 +1788,9 @@ mod tests {
     use leb128;
     use parser::{Error, Format, Result};
     use self::test_assembler::{Endian, Section};
-    use unit::{DebugInfoOffset, UnitOffset};
+    use std::usize;
     use test_util::GimliSectionMethods;
+    use unit::{DebugInfoOffset, UnitOffset};
 
     #[test]
     fn test_compute_pc() {
@@ -1801,7 +1802,7 @@ mod tests {
         assert_eq!(compute_pc(ebuf, ebuf, 0), Ok(*ebuf));
         assert_eq!(
             compute_pc(ebuf, ebuf, -1),
-            Err(Error::BadBranchTarget(-1i64 as u64))
+            Err(Error::BadBranchTarget(usize::MAX as u64))
         );
         assert_eq!(compute_pc(ebuf, ebuf, 5), Ok(ebuf.range_from(5..)));
         assert_eq!(
@@ -2119,7 +2120,7 @@ mod tests {
             let input = [opcode.0, 0xfc, 0xff];
             check_op_parse_failure(
                 &input[..],
-                Error::BadBranchTarget(!0u64),
+                Error::BadBranchTarget(usize::MAX as u64),
                 ADDRESS_SIZE,
                 FORMAT,
             );

--- a/src/parser.rs
+++ b/src/parser.rs
@@ -633,7 +633,7 @@ mod tests {
         let input = [expected.0, 1, 2, 3, 4];
         let input = &mut EndianBuf::new(&input, NativeEndian);
         assert_eq!(parse_pointer_encoding(input), Ok(expected));
-        assert_eq!(*input, EndianBuf::new(&[1, 2, 3, 4], LittleEndian));
+        assert_eq!(*input, EndianBuf::new(&[1, 2, 3, 4], NativeEndian));
     }
 
     #[test]

--- a/src/parser.rs
+++ b/src/parser.rs
@@ -618,6 +618,7 @@ mod tests {
         let section = Section::with_endian(Endian::Little).L64(0x0123456789abcdef);
         let buf = section.get_contents().unwrap();
 
+        let input = &mut EndianBuf::new(&buf, LittleEndian);
         match input.read_offset(Format::Dwarf64) {
             Err(Error::UnsupportedOffset) => assert!(true),
             otherwise => panic!("Unexpected result: {:?}", otherwise),

--- a/src/unit.rs
+++ b/src/unit.rs
@@ -3245,6 +3245,35 @@ mod tests {
     }
 
     #[test]
+    fn test_parse_type_unit_header_32_ok() {
+        let expected_rest = &[1, 2, 3, 4, 5, 6, 7, 8, 9];
+        let mut expected_unit = TypeUnitHeader {
+            header: UnitHeader {
+                unit_length: 0,
+                version: 4,
+                debug_abbrev_offset: DebugAbbrevOffset(0x08070605),
+                address_size: 8,
+                format: Format::Dwarf32,
+                entries_buf: EndianBuf::new(expected_rest, LittleEndian),
+            },
+            offset: DebugTypesOffset(0),
+            type_signature: DebugTypeSignature(0xdeadbeefdeadbeef),
+            type_offset: UnitOffset(0x78563412),
+        };
+        let section = Section::with_endian(Endian::Little)
+            .type_unit(&mut expected_unit)
+            .append_bytes(expected_rest);
+        let buf = section.get_contents().unwrap();
+        let rest = &mut EndianBuf::new(&buf, LittleEndian);
+
+        assert_eq!(
+            parse_type_unit_header(rest, DebugTypesOffset(0)),
+            Ok(expected_unit)
+        );
+        assert_eq!(*rest, EndianBuf::new(expected_rest, LittleEndian));
+    }
+
+    #[test]
     #[cfg(target_pointer_width = "64")]
     fn test_parse_type_unit_header_64_ok() {
         let expected_rest = &[1, 2, 3, 4, 5, 6, 7, 8, 9];
@@ -3284,7 +3313,6 @@ mod tests {
     }
 
     #[test]
-    #[cfg(target_pointer_width = "64")]
     fn test_attribute_value() {
         let mut unit = test_parse_attribute_unit_default();
         let endian = unit.entries_buf.endian();
@@ -3324,6 +3352,7 @@ mod tests {
                 AttributeValue::Data4(([4, 3, 2, 1], endian)),
                 AttributeValue::Udata(0x01020304),
             ),
+            #[cfg(target_pointer_width = "64")]
             (
                 2,
                 constants::DW_AT_data_member_location,

